internal/pkg/nftTrade-scrapers/opensea.go

```diff
@@ -0,0 +1,454 @@
+
+			return err
+		}
+
+		// store the log index of last processed event
+		s.lastLogIndex = &ev.Raw.Index
+	}
+
+	s.lastBlockNum = new(big.Int).SetUint64(endBlockNum + 1)
+	s.lastLogIndex = nil
+
+	return nil
+}
+
+func (s *OpenSeaScraper) notifyTrade(ev *opensea.ContractOrdersMatched, transfer *erc721Transfer) error {
+	nft, err := s.tradeScraper.datastore.GetNFT(transfer.NFTAddress.Hex(), dia.ETHEREUM, transfer.TokenID.String())
+	if err != nil {
+		// TODO: if the error indicates the nft could not found in rdb,
+		//       maybe we can create a new nft record by method SetNFT.
+		//       otherwise, the scraper will stuck at this event due to
+		//       undefined nft. @jppade, what is the desired behaivor?
+		return err
+	}
+
+	{ // TODO: remove the block, it simulates nft exists in the rdb
+		nft = dia.NFT{
+			NFTClass: dia.NFTClass{
+				Address:      transfer.NFTAddress.Hex(),
+				Blockchain:   dia.ETHEREUM,
+				ContractType: "ERC721/ERC1155",
+				Category:     "N/A",
+			},
+			TokenID:        transfer.TokenID.String(),
+			CreationTime:   time.Now(),
+			CreatorAddress: "N/A", // it can be find it
+			URI:            "N/A", // it can be find it
+			Attributes:     nil,   // TODO ??
+		}
+
+		if transfer.Name != nil {
+			nft.NFTClass.Name = *transfer.Name
+		}
+
+		if transfer.Symbol != nil {
+			nft.NFTClass.Symbol = *transfer.Symbol
+		}
+	}
+
+	trade := dia.NFTTrade{
+		NFT:         nft,
+		BlockNumber: new(big.Int).SetUint64(ev.Raw.BlockNumber),
+		// different NFTs can use different asset instead of eth for trades,
+		// it could be found the used asset token by checking more log records
+		// for now we assume eth is used for trades
+		PriceUSD:    s.calcUSDFromWEI(ev.Price, ev.Raw.BlockNumber),
+		FromAddress: transfer.From,
+		ToAddress:   transfer.To,
+		Exchange:    "", // TODO: @jppade, what should its value be? may be `OpenSea`?
+		TxHash:      ev.Raw.TxHash,
+	}
+
+	// handle close request too if the chanTrade not consumed immediately
+	select {
+	case s.tradeScraper.chanTrade <- trade:
+	case <-s.tradeScraper.shutdown:
+		return errOpenSeaShutdownRequest
+	}
+
+	return nil
+}
+
+func (s *OpenSeaScraper) calcUSDFromWEI(wei *big.Int, blockNum uint64) float64 {
+	ethPrice, err := s.ethPrice(blockNum)
+	if err != nil {
+		return 0
+	}
+
+	eth := decimal.NewFromBigInt(wei, 0).Div(tenToPowerOf18)
+
+	price := eth.Mul(ethPrice)
+
+	// using float type is not a good idea to handle prices
+	// we ignore if the price cannot be presentable as float64
+	f, _ := price.Float64()
+
+	return f
+}
+
+func (s *OpenSeaScraper) ethPrice(blockNum uint64) (decimal.Decimal, error) {
+	// TODO: find the eth price in usd for the given block number
+	return decimal.NewFromString("1")
+}
+
+func (scraper *OpenSeaScraper) FetchTrades() (trades []dia.NFTTrade, err error) {
+	// we are using chanTrade as the trade stream, it supplied periodically.
+	// I don't get it exactly, is this function will be used in separate logic?
+	// TODO: @jppade
```

Comment:
Thanks for the valuable input. As the trades scraping is WIP, this is indeed a remainder of a different conception of the logic. The signature should be 'FetchTrades() error' and chanTrade should be filled inside this function. If you like you can amend it. Otherwise I'll do it later.